In [ ]:
my_list = [10, 3, 4, 7, 8, 9]
result = sum_of_evens(my_list)
print("Sum of even numbers:", result)

Sum of even numbers: 22


In [ ]:
word = "Python"
reversed_word = reverse_string(word)
print("Reversed string:", reversed_word)

Reversed string: nohtyP


In [ ]:
def square_numbers(numbers):
    return [num ** 2 for num in numbers]
my_list = [1, 2, 3, 4, 5]
result = square_numbers(my_list)
print("Squared numbers:", result)




Squared numbers: [1, 4, 9, 16, 25]


In [ ]:
for num in range(1, 201):
    if is_prime(num):
        print(f"{num} is a prime number.")

2 is a prime number.
3 is a prime number.
5 is a prime number.
7 is a prime number.
11 is a prime number.
13 is a prime number.
17 is a prime number.
19 is a prime number.
23 is a prime number.
29 is a prime number.
31 is a prime number.
37 is a prime number.
41 is a prime number.
43 is a prime number.
47 is a prime number.
53 is a prime number.
59 is a prime number.
61 is a prime number.
67 is a prime number.
71 is a prime number.
73 is a prime number.
79 is a prime number.
83 is a prime number.
89 is a prime number.
97 is a prime number.
101 is a prime number.
103 is a prime number.
107 is a prime number.
109 is a prime number.
113 is a prime number.
127 is a prime number.
131 is a prime number.
137 is a prime number.
139 is a prime number.
149 is a prime number.
151 is a prime number.
157 is a prime number.
163 is a prime number.
167 is a prime number.
173 is a prime number.
179 is a prime number.
181 is a prime number.
191 is a prime number.
193 is a prime number.
197 is a prime nu

In [ ]:

fib = Fibonacci(10)

for number in fib:
    print(number)

0
1
1
1
2
3
5
8
13
21


In [ ]:
for value in powers_of_two(5):
    print(value)





1
2
4
8
16
32


In [5]:
def read_file_line_by_line(file_path):
    """
    Generator function that reads a file line by line.

    :param file_path: Path to the file
    :yield: Each line in the file as a string (including newline character)
    """
    with open(file_path, 'r') as file:
        for line in file:
            yield line  # You can use line.rstrip('\n') to remove newline if preferred

In [ ]:
data = [(1, 3), (4, 1), (2, 5), (3, 2)]
sorted_data = sorted(data, key=lambda x: x[1])
print("Sorted list:", sorted_data)

Sorted list: [(4, 1), (3, 2), (1, 3), (2, 5)]


In [ ]:

celsius_temps = [0, 20, 37, 100]
fahrenheit_temps = list(map(lambda c: (c * 9/5) + 32, celsius_temps))
print("Temperatures in Fahrenheit:", fahrenheit_temps)

Temperatures in Fahrenheit: [32.0, 68.0, 98.6, 212.0]


In [ ]:
def remove_vowels(text):
    vowels = "aeiouAEIOU"
    return ''.join(filter(lambda char: char not in vowels, text))
input_string = "Hello, World!"
result = remove_vowels(input_string)
print("String without vowels:", result)

String without vowels: Hll, Wrld!


In [ ]:

orders = [
    [34587, "Learning Python, Mark Lutz", 4, 40.95],
    [98762, "Programming Python, Mark Lutz", 5, 56.80],
    [77226, "Head First Python, Paul Barry", 3, 32.95],
    [88112, "Einführung in Python3, Bernd Klein", 3, 24.99]
]
final_orders = list(map(
    lambda order: (
        order[0],
        round(order[2] * order[3] + (10 if order[2] * order[3] < 100 else 0), 2)
    ),
    orders
))
print("Processed Orders:")
for item in final_orders:
    print(item)

Processed Orders:
(34587, 163.8)
(98762, 284.0)
(77226, 108.85)
(88112, 84.97)


1. What is the difference between a function and a method in Python?

In Python, both functions and methods are blocks of reusable code that perform a specific task. However, there are some key differences between them:

 Function
Definition: A function is defined using the def keyword, and it is not associated with an object.

Can be called independently.

Defined at the module level or inside another function.

In [ ]:
def greet(name):
    return f"Hello, {name}!"

print(greet("Kartik"))

Hello, Kartik!


 Method
Definition: A method is a function that belongs to an object (usually a class instance).

It is called on an object, and the object is automatically passed as the first parameter (typically self).

Defined inside a class.

In [ ]:
class Greeter:
    def greet(self, name):
        return f"Hello, {name}!"

g = Greeter()
print(g.greet("Kartik"))

Hello, Kartik!


2. Explain the concept of function arguments and parameters in Python.

The terms arguments and parameters are often used interchangeably, but they have distinct meanings in Python:

 Parameters:
Definition: Variables listed inside the parentheses in the function definition.

These are placeholders for the values the function will receive when it is called.

In [ ]:
def greet(name, message):
    print(f"{message}, {name}!")

 Arguments:
Definition: Actual values you pass to a function when calling it.

These get assigned to the corresponding parameters.

In [ ]:
greet("Kartik", "Good Morning")  # "Kartik" and "Good Morning" are arguments

Good Morning, Kartik!


3. What are the different ways to define and call a function in Python?

In Python, functions can be defined and called in multiple ways, depending on the need and flexibility required. Here’s a breakdown:

Standard Function Definition
Using the def keyword:

In [ ]:
def greet(name):
    return f"Hello, {name}!"

Function with Variable-Length Arguments
*args: for variable number of positional arguments

**kwargs: for variable number of keyword arguments

In [ ]:
def show_args(*args, **kwargs):
    print("Args:", args)
    print("Kwargs:", kwargs)

Ways to Call a Function
1. Positional Arguments

In [ ]:
greet("Kartik")

'Hello, Kartik!'

Keyword Arguments

4. What is the purpose of the `return` statement in a Python function?

The return statement in Python is used to send a value back to the caller of a function. It ends the function execution and passes the specified value back to where the function was called.



In [ ]:
def add(a, b):
    return a + b

result = add(3, 5)
print(result)  # Output: 8

8


5. What are iterators in Python and how do they differ from iterables?

In Python, iterables and iterators are both used to loop through data, but they are not the same.

An iterable is any object you can loop over with a for loop.

It implements the __iter__() method and returns an iterator.

Examples: list, tuple, str, dict, set, etc.

python
Copy
Edit


In [ ]:
my_list = [1, 2, 3]
for item in my_list:  # my_list is iterable
    print(item)

1
2
3


An iterator is an object that remembers its position during iteration.

It has two main methods:

__iter__() → returns the iterator itself

__next__() → returns the next item or raises StopIteration

In [ ]:
my_list = [1, 2, 3]
it = iter(my_list)   # create an iterator from the iterable

print(next(it))  # 1
print(next(it))  # 2
print(next(it))  # 3
# print(next(it))  # Raises StopIteration

1
2
3


6. Explain the concept of generators in Python and how they are defined.

Generators in Python are a special type of iterator that allow you to generate values on the fly, instead of storing them all in memory. They are useful for working with large data, improving performance and efficiency.

A generator is:

A function that uses the yield keyword instead of return

It returns a generator object that produces items one at a time

It remembers its state between each call

How to Define a Generator
Using a Generator Function:

In [ ]:
def count_up_to(n):
    count = 1
    while count <= n:
        yield count  # pauses and returns value
        count += 1

Advantages of Generators:
Efficient memory usage (no need to store large sequences)

Faster for large datasets

Can be infinite (like reading from a stream)

7. What are the advantages of using generators over regular functions?

Generators provide several benefits, especially when dealing with large data, streams, or performance-critical applications.

op Advantages of Generators
1. Memory Efficiency
Regular functions build and return the entire data structure (e.g., list), which uses more memory.

Generators yield one item at a time, don’t store all values in memory.

In [ ]:
def gen_nums():
    for i in range(1000000):
        yield i

Lazy Evaluation
Generators evaluate items only when needed using next().

This is perfect for large/infinite sequences, like reading from a file or API.

In [ ]:
def infinite_counter():
    n = 1
    while True:
        yield n
        n += 1

8. What is a lambda function in Python and when is it typically used?

A lambda function in Python is a small, anonymous function defined using the lambda keyword. It is typically used for short, simple operations, especially where defining a full def function would be overkill.



In [ ]:
square = lambda x: x * x
print(square(5))  # Output: 25

25


Common Use Cases of Lambda Functions:
 Used with map()

In [ ]:
nums = [1, 2, 3]
squares = list(map(lambda x: x ** 2, nums))  # [1, 4, 9]

Used with filter()
Filter items in a list based on a condition:

In [ ]:
nums = [1, 2, 3, 4, 5]
evens = list(filter(lambda x: x % 2 == 0, nums))  # [2, 4]

9. Explain the purpose and usage of the `map()` function in Python.

The map() function in Python is used to apply a function to every item in an iterable (like a list, tuple, etc.) and return a new iterator with the results.

 Purpose of map()
To perform transformation or processing on each item in a sequence

Avoids writing explicit for loops

Makes code shorter, more readable, and efficient



In [ ]:
def square(x):
    return x * x

nums = [1, 2, 3, 4]
squared = map(square, nums)
print(list(squared))  # Output: [1, 4, 9, 16]

[1, 4, 9, 16]


10. What is the difference between `map()`, `reduce()`, and `filter()` functions in Python?

All three — map(), filter(), and reduce() — are functional programming tools in Python that work with iterables. They help apply functions across elements of sequences like lists or tuples but serve different purposes.

 map() → Transforms Each Item
Purpose: Applies a function to every item in an iterable and returns an iterator with the results

In [ ]:
nums = [1, 2, 3]
squares = map(lambda x: x**2, nums)
print(list(squares))  # Output: [1, 4, 9]

[1, 4, 9]


filter() → Filters Items Based on Condition
Purpose: Applies a function that returns True or False to each item and filters only the items that return True.

In [ ]:
nums = [1, 2, 3, 4, 5]
evens = filter(lambda x: x % 2 == 0, nums)
print(list(evens))  # Output: [2, 4]

[2, 4]


reduce() → Reduces Items to a Single Value
Purpose: Repeatedly applies a function to pairs of items in a sequence, reducing it to a single value.

Note: reduce() is not built-in; it must be imported from functools.

In [ ]:
from functools import reduce

nums = [1, 2, 3, 4]
product = reduce(lambda x, y: x * y, nums)
print(product)  # Output: 24

24


11 Using pen & Paper write the internal mechanism for sum operation using  reduce function on this given
list:[47,11,42,13];

the internal mechanism of the reduce() function step-by-step by hand (pen & paper style) for the list:

In [ ]:
[47, 11, 42, 13]

[47, 11, 42, 13]

[47, 11, 42, 13]:
Let func = lambda x, y: x + y

Step 1:

x = 47, y = 11
result1 = 47 + 11 = 58



Step 2:

x = 58, y = 42
result2 = 58 + 42 = 100



Step 3:

x = 100, y = 13
result3 = 100 + 13 = 113
